File: ios-app/src/services/temerantRpg.ts
import apiClient from './api';

export interface TemerantRpgCharacter {
  id: string;
  user_id: string;
  character_name: string;
  origin?: string | null;
  backstory?: string | null;
  body: number;
  mind: number;
  craft: number;
  voice: number;
  luck: number;
  coin_talents: number;
  rank: string;
  conditions: Record<string, any>;
  skills: Record<string, number>;
  inventory: string[];
  term_index: number;
  current_scene_id?: string | null;
}

export interface TemerantRpgWorldState {
  local_date: string;
  day_slot: string;
  weather: string;
  location_hint: string;
  ambient_events: string[];
  pending_consequences: Array<Record<string, any>>;
  last_advance_summary?: string | null;
}

export interface TemerantRpgScene {
  id: string;
  scene_number: number;
  local_date: string;
  day_slot: string;
  location: string;
  title: string;
  opening_text: string;
  status: 'open' | 'closed';
  summary?: string | null;
  consequences: Array<Record<string, any>>;
  opened_at: string;
  closed_at?: string | null;
}

export interface TemerantRpgRelationship {
  npc_key: string;
  display_name: string;
  disposition: string;
  trust: string;
  respect: string;
  debt_balance: number;
  notes?: string | null;
}

export interface TemerantRpgState {
  character: TemerantRpgCharacter;
  world: TemerantRpgWorldState;
  open_scene?: TemerantRpgScene | null;
  relationships: TemerantRpgRelationship[];
}

export interface TemerantRpgTurnResponse {
  scene_id: string;
  turn_index: number;
  outcome: string;
  total: number;
  difficulty: number;
  margin: number;
  response_text: string;
  consequence?: Record<string, any> | null;
}

export interface TemerantRpgJournalEntry {
  id: string;
  local_date: string;
  summary_markdown: string;
  scene_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface TemerantRpgTerm {
  id: string;
  term_index: number;
  month: string;
  admissions_result: string;
  tuition_talents: number;
  summary: string;
  created_at: string;
  updated_at: string;
}

class TemerantRpgService {
  async createCharacter(payload: {
    character_name: string;
    origin?: string;
    backstory?: string;
  }): Promise<TemerantRpgCharacter> {
    return apiClient.post('/api/temerant-rpg/characters', payload);
  }

  async getState(): Promise<TemerantRpgState> {
    return apiClient.get('/api/temerant-rpg/state');
  }

  async openScene(payload?: {
    title?: string;
    location?: string;
    opening_prompt?: string;
  }): Promise<TemerantRpgScene> {
    return apiClient.post('/api/temerant-rpg/scenes/open', payload || {});
  }

  async actInScene(sceneId: string, payload: {
    action: string;
    attribute?: string;
    skill?: string;
  }): Promise<TemerantRpgTurnResponse> {
    return apiClient.post(`/api/temerant-rpg/scenes/${sceneId}/act`, payload);
  }

  async closeScene(sceneId: string, summary?: string): Promise<TemerantRpgScene> {
    return apiClient.post(`/api/temerant-rpg/scenes/${sceneId}/close`, { summary });
  }

  async advanceTime(slots: number = 1): Promise<{ local_date: string; day_slot: string; summary: string }> {
    return apiClient.post('/api/temerant-rpg/time/advance', { slots });
  }

  async listJournal(limit: number = 10): Promise<TemerantRpgJournalEntry[]> {
    return apiClient.get(`/api/temerant-rpg/journal?limit=${limit}`);
  }

  async runAdmissions(term_index?: number): Promise<TemerantRpgTerm> {
    return apiClient.post('/api/temerant-rpg/admissions/run', { term_index });
  }

  async listTerms(limit: number = 6): Promise<TemerantRpgTerm[]> {
    return apiClient.get(`/api/temerant-rpg/terms?limit=${limit}`);
  }
}

export const temerantRpgService = new TemerantRpgService();
export default temerantRpgService;

File: backend/app/schemas/temerant_rpg.py
"""Schemas for the separate scene-based Temerant RPG API."""

from datetime import date, datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class TemerantRpgCharacterCreate(BaseModel):
    character_name: str = Field(default="Daveth of Andentown", min_length=1, max_length=120)
    origin: Optional[str] = None
    backstory: Optional[str] = None


class TemerantRpgCharacterResponse(BaseModel):
    id: str
    user_id: str
    character_name: str
    origin: Optional[str] = None
    backstory: Optional[str] = None
    body: int
    mind: int
    craft: int
    voice: int
    luck: int
    coin_talents: float
    rank: str
    conditions: Dict[str, Any]
    skills: Dict[str, int]
    inventory: List[str]
    term_index: int
    current_scene_id: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TemerantRpgWorldStateResponse(BaseModel):
    local_date: date
    day_slot: str
    weather: str
    location_hint: str
    ambient_events: List[str]
    pending_consequences: List[Dict[str, Any]]
    last_advance_summary: Optional[str] = None


class TemerantRpgRelationshipResponse(BaseModel):
    npc_key: str
    display_name: str
    disposition: str
    trust: str
    respect: str
    debt_balance: int
    notes: Optional[str] = None


class TemerantRpgSceneOpenRequest(BaseModel):
    title: Optional[str] = None
    location: Optional[str] = None
    opening_prompt: Optional[str] = None


class TemerantRpgSceneResponse(BaseModel):
    id: str
    scene_number: int
    local_date: date
    day_slot: str
    location: str
    title: str
    opening_text: str
    status: str
    summary: Optional[str] = None
    consequences: List[Dict[str, Any]]
    opened_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None


class TemerantRpgSceneTurnRequest(BaseModel):
    action: str = Field(..., min_length=1, max_length=2000)
    attribute: Optional[str] = Field(default=None, description="body|mind|craft|voice|luck")
    skill: Optional[str] = Field(default=None, max_length=64)
    difficulty: Optional[int] = Field(default=None, ge=3, le=30)
    circumstance_mod: int = Field(default=0, ge=-6, le=6)


class TemerantRpgSceneTurnResponse(BaseModel):
    scene_id: str
    turn_index: int
    outcome: str
    total: int
    difficulty: int
    margin: int
    response_text: str
    consequence: Optional[Dict[str, Any]] = None


class TemerantRpgCloseSceneRequest(BaseModel):
    summary: Optional[str] = None


class TemerantRpgAdvanceTimeRequest(BaseModel):
    slots: int = Field(default=1, ge=1, le=9)


class TemerantRpgAdvanceTimeResponse(BaseModel):
    local_date: date
    day_slot: str
    summary: str


class TemerantRpgStateResponse(BaseModel):
    character: TemerantRpgCharacterResponse
    world: TemerantRpgWorldStateResponse
    open_scene: Optional[TemerantRpgSceneResponse] = None
    relationships: List[TemerantRpgRelationshipResponse]


class TemerantRpgJournalEntryResponse(BaseModel):
    id: str
    local_date: date
    summary_markdown: str
    scene_ids: List[str]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TemerantRpgGenerateJournalRequest(BaseModel):
    local_date: Optional[date] = None
    regenerate: bool = True


class TemerantRpgTermResponse(BaseModel):
    id: str
    term_index: int
    month: date
    admissions_result: str
    tuition_talents: float
    summary: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TemerantRpgRunAdmissionsRequest(BaseModel):
    term_index: Optional[int] = None
